### PR TITLE
Deprecate BoltDB, preventing creation of new databases.

### PR DESCRIPTION
**Pull request by mheon**
This is one of the breaking changes in Podman 5.0: removing the ability to create new instances of the old Bolt database. This does not remove support for the database entirely, as existing Bolt databases will still be usable, but all new installs will use SQLite after this point - if Bolt is forced by config, we'll just error.

We don't have plans to outright remove the Bolt code. If that were to happen, it'd be Podman 6.0 at least, and a significant enough change it'd warrant a lot of discussion and planning. We do intend to start winding down support of BoltDB, though, and new features may be added only to SQLite from here on.

I have added an escape hatch via an undocumented environment variable that allows us to continue testing BoltDB in CI (and, if necessary, locally) but I don't want this to be used for any purpose except continued testing of the old DB to ensure we don't break it.

#### Does this PR introduce a user-facing change?

```release-note
It is no longer possible to create new BoltDB databases; attempting to do so will result in an error. All new Podman installations will now use the SQLite database backend.
```

### CODE REVIEW

**Comment by mheon** [comment]
@edsantiago @cevich PTAL

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:118]: As an FYI (I don't know it to be a problem yet): Adding any differentiation element to a matrix item will result in Cirrus forcing a unique (and often unreadable) task name.  It only does this if `name` is not already unique among the matrix.  That's why we have the `&std_name_fmt` YAML alias.  It (previously) results in mostly human-readable task names.

So, it may be worth considering if adding `$ALLOW_LEGACY_DB` to that alias might be helpful to people trying to sift through 91-some-odd checks.

**Review by cevich — COMMENTED** [review]
Upon initial (quick) inspection, this looks good.  I haven't taken a look at the test results (or names), leaving that up to a future review and/or reviewer.

File: .cirrus.yml
Comment (mheon) [reply .cirrus.yml:118]: It should be set everywhere CI_DESIRED_DATABASE is set to "boltdb".  I mostly wanted a new variable for readability.

Comment (cevich) [reply .cirrus.yml:118]: Just looked through a fistful of tasks.  You're right, it seems to be okay. The Cirrus-CI Name-uniqueness `$DIETY` is fully satisfied :grin:

**Comment by rhatdan** [comment]
Can we add a test to set the environment to false containers.conf to sqlite and make sure podman fails with no sqlite database?

**Comment by mheon** [comment]
Now that I think about it, that ought to be possible

File: test/system/001-basic.bats
Comment (mheon) [inline test/system/001-basic.bats:284]: @edsantiago I'm pretty sure I'm doing this wrong

File: .cirrus.yml
Comment (edsantiago) [inline .cirrus.yml:64]: I'm obviously missing something, but I just can't understand what I'm missing.

Why is `$CI_DESIRED_DATABASE` not sufficient?

File: test/system/001-basic.bats
Comment (edsantiago) [reply test/system/001-basic.bats:284]: Hoo boy. Coupla things.

First: `$CI_DESIRED_DATABASE` does not tell _podman_ to use boltdb/sqlite, it tell *our Cirrus setup* to do magic setup. So this test is a NOP.

Second, and much more importantly, the way to test this would be `--db-backend` and please allow me to remind you of #20563. If any test ever does `--db-backend=boltdb`, podman is forever hosed, not even `system reset` will fix things. 

This is going to be really hard to test. My recommendation:
1. Revert your `.cirrus.yml` changes 
2. Keep your Go changes, but make the bypass be `$CI_DESIRED_DATABASE`. Stick with what works.
3. Revert the system test. I think this has to be done in an e2e test. If it has to be done in system tests, it might be necessary to use `--runroot --root --this --that --make --sure --you --use --every --tmpdir --available`

But then again I might be missing something (See above).

**Review by edsantiago — CHANGES_REQUESTED** [review]
I think we can do this more cleanly. Testing, though ... testing is hard.

**Comment by mheon** [comment]
I was a little reluctant to incorporate CI details into the PR, but I don't think it's harmful now that I think about it more. Probably easier, even.

**Comment by mheon** [comment]
I think the `system reset` bit of #20563 is fixed (it's definitely not intended behavior, and if it's not fixed we need to fix it). So, theoretically, we could swap DB in system tests so long as it was used?

File: libpod/boltdb_state.go
Comment (edsantiago) [inline libpod/boltdb_state.go:91]: Suggestion: flip the conditions (test exist first, and then env). Abort or warn:
```go
    if ... Stat, ErrNotExist {
        if ($CI_DESIRED_DATABASE==boltdb) {
            warn("BoltDB is deprecated; proceeding due to $CI_DESIRED_DATABASE. This is your only warning.")
        } else {
            return nil, Error(...)
        }
    }
```

File: test/system/001-basic.bats
Comment (edsantiago) [reply test/system/001-basic.bats:284]: Test suggestion:
```bash
@test "podman - BoltDB cannot create new databases" {
    safe_opts=$(podman_isolation_opts ${PODMAN_TMPDIR})

    CI_DESIRED_DATABASE= run_podman 125 $safe_opts --db-backend=boltdb info
    assert "$output" =~ "deprecated, no new BoltDB databases can be created" \
           "without CI_DESIRED_DATABASE"

    CI_DESIRED_DATABASE=boltdb run_podman $safe_opts --db-backend=boltdb info
    assert "$output" =~ "deprecated; proceeding due to" \
           "with CI_DESIRED_DATABASE"
}
```
You have to add this to `005-info.bats` (or elsewhere), not `001`, because `001` is super paranoid safe and so it does not load `helpers.bash` and so it does not get `podman_isolation_opts`.

This of course assumes that you accept my suggestions above. Modify as needed.

File: libpod/boltdb_state.go
Comment (mheon) [reply libpod/boltdb_state.go:91]: Can't really drop a warning here, unfortunately... We init Libpod multiple times per run on rootless, so we'd be spamming folks with multiple messages per run. I don't think there's an easy way around that, unfortunately, we don't really share anything between the runs.

Maybe a debug log, since those are already overly-verbose.

Comment (vrothberg) [reply libpod/boltdb_state.go:91]: There's still the hidden `--db-backend` flag.  Could we just continue keeping that around?  The new env variable confuses me as it adds (yet) another way to configure/set the DB.

File: test/system/001-basic.bats
Comment (mheon) [reply test/system/001-basic.bats:284]: @edsantiago Thanks bunches. Added a `system reset` after to try and purge the remains of the tmpdir Podman, which I think ought to be safe this early in the test suite

**Comment by TomSweeneyRedHat** [comment]
If you can make the tests happy, LGTM

**Comment by edsantiago** [comment]
As an end user, I'm not 100% sure I would be happy if I upgrade and all of a sudden nothing works at all.

**Comment by mheon** [comment]
Yeah, I don't think we can remove it without a migration strategy, which we don't have time to do at this point. Maybe for 6.0.

File: test/system/005-info.bats
Comment (edsantiago) [inline test/system/005-info.bats:258]: Needs `skip_if_remote`. Sorry for not realizing that earlier.

**Review by edsantiago — CHANGES_REQUESTED** [review]
And I think `info_test.go:It("podman --db-backend info basic check")` needs major surgery, possibly resection.

**Comment by cevich** [comment]
I don't mean to belabor my point, but couldn't "migration strategy" be fairly simple documentation:  Export your volumes, secrets, and network setup.  Copy to new install of podma (different machine) or safely wipe existing + new setup w/ SQLite.  Re-import volumes, secrets, and network setup.  Done.

No?

(My intention here is to try and simplify our life.  Since whatever's in 5.0 CI will stick around fora potentially VERY long time.  Let's make our life easier if we can)

**Comment by mheon** [comment]
That might work for Fedora, but this has to go into RHEL 9, so we have to keep Bolt around for the lifespan of RHEL9 feature releases.

**Comment by cevich** [comment]
"this has to go into RHEL 9" - Oh right, damn.

`/me` thought he remembered something about NOT doing major version bumps in RHEL like this.  I think the need for this PR illustrates why :disappointed:  Grumble...grumble...grumble...okay I'll shut my mouth now :man_shrugging:

File: test/e2e/info_test.go
Comment (edsantiago) [inline test/e2e/info_test.go:210]: I'm really uncomfortable with this: if one of the `Expect()s` below fails, it'll leave the envariable set, and subsequent tests will fail in distracting/redherring ways. Would it be possible to
```
     const desiredDB ...
+    oldDB = os.Getenv()
+    defer { restore oldDB }
```
Way up there instead of here?

File: test/e2e/info_test.go
Comment (edsantiago) [inline test/e2e/info_test.go:220]: Why the `if`?

File: test/e2e/info_test.go
Comment (edsantiago) [inline test/e2e/info_test.go:205]: May I ask for one more case here? Append:
```
    // just because we requested sqlite doesn't mean it stays that way. 
    // once a boltdb exists, podman will forevermore stick with it
    arg: "", want: "boltdb"},
```

File: test/e2e/info_test.go
Comment (mheon) [reply test/e2e/info_test.go:210]: We'd have to retain the existing cleanup code as that handles it on a per-loop basis not a per-function basis but we can definitely add the defer in addition, it shouldn't hurt even if cleanup was otherwise successful.

File: test/e2e/info_test.go
Comment (mheon) [reply test/e2e/info_test.go:205]: Sure

File: libpod/state_test.go
Comment (edsantiago) [inline libpod/state_test.go:49]: Is there any cleanup needed here?

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
Well, [yuk](https://api.cirrus-ci.com/v1/artifact/task/5393091192946688/html/int-podman-debian-13-rootless-host-sqlite.log.html#t--Podman-Info-podman-db-backend-info-basic-check--1). It's because of the way e2e tests are set up, where every podman command has an enormous `--this --that --whatever` set of explicit options. One of those is `--db-backend`, so the command that's being run is actually:
```
$ podman [options] --log-level=info info --format {{.Host.DatabaseBackend}}
          ^^^^^^^-> which includes --db-backend sqlite
```
...whereas before it was
```
--db-backend sqlite [crap crap] --db-backend ""
```
...which I guess is not the same thing.

I still think deleting this entire test is the best thing to do, but if you want to put back the `""` code I guess that'll work. It's just not really a test of anything.

**Comment by edsantiago** [comment]
Weird. in [f38 rootless](https://api.cirrus-ci.com/v1/artifact/task/5393550720892928/html/sys-podman-fedora-38-root-host-boltdb.log.html#t--00273):
```
<+056ms> # # podman image scp foo.bar/nonesuch/c_iypxcy1e6m:mytag some25515dude@localhost::
...
         # Error: the BoltDB backend has been deprecated, no new BoltDB databases can be created: invalid argument
```
This test podman-scps an image from root to rootless or vice-versa. I can't figure out where the destination is getting the idea that it should initialize as boltdb. I will look closer tomorrow.

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM once @edsantiago is happy with the tests.  At the moment, the CI is hiccuping, I'm not sure if it's related.

**Comment by cevich** [comment]
> where the destination is getting the idea that it should initialize as boltdb. I will look closer tomorrow.

Ed, I haven't been watching this PR very closely.  But if the "idea" is passed around in an env. var., then care should be taken to ensure it's mentioned in [one of the special lists](https://github.com/containers/podman/blob/main/contrib/cirrus/lib.sh#L98-L104) (example).

**Comment by edsantiago** [comment]
Yep, close, but I don't think it's the list thing. And it's not `/etc/cirrus_environment` because the desired envariable is already there. I need to figure out something such that:
```
# ssh rootless@localhost cmd
```
...loads the desired environment. Maybe mucking with the user's .bashrc, or system bashrc, I dunno, almost meeting time so I'm giving up for now.

**Comment by cevich** [comment]
> loads the desired environment

[ssh re-exec's](https://github.com/containers/podman/blob/main/contrib/cirrus/runner.sh#L527-L529) the `runner.sh` script which loads `lib.sh`

**Comment by cevich** [comment]
`lib.sh` then [loads `ci_environment`](https://github.com/containers/podman/blob/main/contrib/cirrus/lib.sh#L36-L37)

**Comment by cevich** [comment]
To clarify: `/etc/cirrus_environment` != `/etc/ci_environment` != `$CIRRUS_ENV`.  `/etc/ci_environment` is the preferred mechanism (in this CI) for making variables persist across execution contexts.

**Comment by edsantiago** [comment]
Somehow podman is invoking ssh in a way that the remote end does not source `rc` files. I can't figure how it does this, and can't reproduce a command-line-ssh way to say "don't source rc files". 

Anyhow, giving up. I suggest just
```bash
    if [[ "$CI_DESIRED_DATABASE" = "boltdb" ]]; then
        skip "this is a nightmare"
    fi
```

**Review by edsantiago — APPROVED** [review]
Thank you @mheon for persevering.

This LGTM, and I love how clean the code is (the actual code, not the tests!). I will not slash-lgtm because I am too fried to give this a final thorough review ... but a definite slashless LGTM.

**Comment by rhatdan** [comment]
/hold

@mheon Looks like you have a lot of don't merge tags?

**Comment by edsantiago** [comment]
Blush. That was me, when I was playing with the live CI VM and thought the job might pass. It would've been very very bad if that had merged. Sorry for forgetting to clear those. I've cleared them, and mergebot should notice soon.

**Pull request merged**
